src/descriptors.rs

```diff
@@ -180,6 +182,153 @@ macro_rules! descriptor_fields {
+    /// one from your own descriptor bytes for tests.
+    ///
+    /// ### Panics
+    ///  * when the buffer is too short for a device descriptor
+    ///  * when the first descriptor is not a device descriptor
+    pub fn new(buf: &'a [u8]) -> Self {
+        assert!(buf.len() >= DESCRIPTOR_LEN_DEVICE as usize);
+        assert!(buf[0] as usize >= DESCRIPTOR_LEN_DEVICE as usize);
+        assert!(buf[1] == DESCRIPTOR_TYPE_DEVICE);
+        Self(buf)
+    }
+
+    /// Get speed of this device. The speed is judged by `usb_version` field.
+    pub fn speed(&self) -> Option<Speed> {
+        Speed::from_usb_version(self.usb_version())
+    }
+}
+
+descriptor_fields! {
+    impl<'a> DeviceDescriptor<'a> {
+        /// `bcdUSB` descriptor field: USB Specification Number.
+        #[doc(alias = "bcdUSB")]
+        pub fn usb_version at 2 -> u16;
+
+        /// `bDeviceClass` descriptor field: Class code, assigned by USB-IF.
+        #[doc(alias = "bDeviceClass")]
+        pub fn class at 4 -> u8;
+
+        /// `bDeviceSubClass` descriptor field: Subclass code, assigned by USB-IF.
+        #[doc(alias = "bDeviceSubClass")]
+        pub fn subclass at 5 -> u8;
+
+        /// `bDeviceProtocol` descriptor field: Protocol code, assigned by USB-IF.
+        #[doc(alias = "bDeviceProtocol")]
+        pub fn protocol at 6 -> u8;
+
+        /// `bMaxPacketSize0` descriptor field: Maximum packet size for 0 Endpoint.
+        #[doc(alias = "bMaxPacketSize0")]
+        pub fn max_packet_size_0 at 7 -> u8;
+
+        /// `idVendor` descriptor field: Vendor ID, assigned by USB-IF.
+        #[doc(alias = "idVendor")]
+        pub fn vendor_id at 8 -> u16;
+
+        /// `idProduct` descriptor field: Product ID, assigned by the manufacturer.
+        #[doc(alias = "idProduct")]
+        pub fn product_id at 10 -> u16;
+
+        /// `bcdDevice` descriptor field: Device release number.
+        #[doc(alias = "bcdDevice")]
+        pub fn device_version at 12 -> u16;
+
+        /// `iManufacturer` descriptor field: Index for manufacturer description string.
+        pub fn manufacturer_string_index at 14 -> u8;
+
+        /// `iProduct` descriptor field: Index for product description string.
+        pub fn product_string_index at 15 -> u8;
+
+        /// `iSerialNumber` descriptor field: Index for serial number string.
+        pub fn serial_number_string_index at 16 -> u8;
+
+        /// `bNumConfigurations` descriptor field: Number of configurations
+        #[doc(alias = "bNumConfigurations")]
+        pub fn num_configurations at 17 -> u8;
+    }
+}
+
+fn format_usb_version(ver: u16) -> &'static str {
+    match ver {
+        0x0100 => "1.0",
+        0x0110 => "1.1",
+        0x0200 => "2.0",
+        0x0210 => "2.0", // This means a USB 3.x device is running in USB 2.0 mode
+        0x0300 => "3.0",
+        0x0310 => "3.1",
+        0x0320 => "3.2",
+        _ => "unknown",
```

Comment:
I don't think this formatting is worth it. In a `Debug` format, it's for developers using the API, so they'll want to see the actual value in hex.